core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java

```diff
@@ -239,77 +240,153 @@
+     * @param type a type of the given {@link Parameter} or {@link Field}. It is a type of the specified
+     *             {@code typeElement} parameter.
+     * @param pathParams a set of path variables.
+     * @param converters a list of {@link RequestConverterFunction} to be used for generating an object.
+     *                   One of the resolver in the list would create the object to be injected.
+     * @param applyConverterForNoAnnotation a flag which indicates how to handle an element which does not have
+     *                                      any annotation. If it is {@code true}, a list of
+     *                                      {@link RequestConverterFunction} would be applied for creating an
+     *                                      object for the element. Otherwise, the element would be ignored.
      */
     private static Optional<AnnotatedValueResolver> of(AnnotatedElement annotatedElement,
                                                        AnnotatedElement typeElement, Class<?> type,
                                                        Set<String> pathParams,
-                                                       List<RequestObjectResolver> objectResolvers) {
+                                                       List<RequestConverterFunction> converters,
+                                                       boolean applyConverterForNoAnnotation) {
         requireNonNull(annotatedElement, "annotatedElement");
         requireNonNull(typeElement, "typeElement");
         requireNonNull(type, "type");
         requireNonNull(pathParams, "pathParams");
-        requireNonNull(objectResolvers, "objectResolvers");
+        requireNonNull(converters, "converters");
 
         final Param param = annotatedElement.getAnnotation(Param.class);
         if (param != null) {
             final String name = findName(param, typeElement);
             if (pathParams.contains(name)) {
-                return Optional.of(ofPathVariable(param, name, annotatedElement, typeElement, type));
+                return Optional.of(ofPathVariable(name, annotatedElement, typeElement, type));
             } else {
-                return Optional.of(ofHttpParameter(param, name, annotatedElement, typeElement, type));
+                return Optional.of(ofHttpParameter(name, annotatedElement, typeElement, type));
             }
         }
 
         final Header header = annotatedElement.getAnnotation(Header.class);
         if (header != null) {
-            return Optional.of(ofHeader(header, annotatedElement, typeElement, type));
+            return Optional.of(ofHeader(findName(header, typeElement), annotatedElement, typeElement, type));
         }
 
+        if (type == Optional.class) {
+            // Do not allow Optional<?> for the following auto-injectable types:
+            //  - RequestContext and ServiceRequestContext
+            //  - Request and HttpRequest
```

Comment:
Perhaps we could just inject an `Optional` that's always present, although it's not the most efficient setup? (We could warn about it though.)